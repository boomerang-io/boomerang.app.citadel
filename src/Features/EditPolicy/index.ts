// @ts-expect-error ts-migrate(6142) FIXME: Module './EditPolicy' was resolved to 'C:/Users/Is... Remove this comment to see the full error message
export { default } from "./EditPolicy"