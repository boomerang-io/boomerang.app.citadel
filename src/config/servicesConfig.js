export const BASE_SERVICE_ENV_URL =
  process.env.NODE_ENV === "development"
    ? "http://localhost:8000"
    : window._SERVER_DATA && window._SERVER_DATA.BASE_SERVICE_ENV_URL;

// Standard
export const SERVICE_USERS_BASE_PATH = "/users";
export const SERVICE_USERS_PROFILE_PATH = `${SERVICE_USERS_BASE_PATH}/profile`;
export const SERVICE_USERS_NAVIGATION_PATH = `${SERVICE_USERS_BASE_PATH}/navigation`;

// Product
export const SERVICE_PRODUCT_BASE_PATH = "/citadel";
export const SERVICE_PRODUCT_DEFINITIONS_PATH = `${SERVICE_PRODUCT_BASE_PATH}/definitions`;
export const SERVICE_PRODUCT_INSIGHTS_PATH = `${SERVICE_PRODUCT_BASE_PATH}/policies/insights`;
export const SERVICE_PRODUCT_POLICIES_PATH = `${SERVICE_PRODUCT_BASE_PATH}/policies`;
export const SERVICE_PRODUCT_TEAM_PATH = `${SERVICE_PRODUCT_BASE_PATH}/teams`;
export const SERVICE_PRODUCT_VIOLATIONS_PATH = `${SERVICE_PRODUCT_BASE_PATH}/policies/violations`;

export const SERVICE_REQUEST_STATUSES = {
  FAILURE: "failure",
  SUCCESS: "success"
};
