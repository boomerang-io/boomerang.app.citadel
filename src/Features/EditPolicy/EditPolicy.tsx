import React, { useCallback, useState, useEffect } from "react";
// @ts-expect-error ts-migrate(7016) FIXME: Could not find a declaration file for module 'uuid... Remove this comment to see the full error message
import uuid from "uuid";
import { useQuery, useMutation, queryCache } from "react-query";
import { Helmet } from "react-helmet";
import {
  notify,
  ToastNotification,
  ErrorMessage,
  TextInputSkeleton,
} from "@boomerang-io/carbon-addons-boomerang-react";
import CreateEditPolicyForm from "Components/CreateEditPolicyForm";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'Components/CreateEditPolicyHea... Remove this comment to see the full error message
import CreateEditPolicyHeader from "Components/CreateEditPolicyHeader";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'Components/DefinitionSkeleton'... Remove this comment to see the full error message
import DefinitionSkeleton from "Components/DefinitionSkeleton";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'Config/servicesConfig' or its ... Remove this comment to see the full error message
import { serviceUrl, resolver } from "Config/servicesConfig";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'Config/appConfig' or its corre... Remove this comment to see the full error message
import { appLink } from "Config/appConfig";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'Constants' or its correspondin... Remove this comment to see the full error message
import { POLICY_INTERACTION_TYPES } from "Constants";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module './editPolicy.module.scss' or i... Remove this comment to see the full error message
import styles from "./editPolicy.module.scss";

type Props = {
    history?: any;
    match?: any;
};

function EditPolicy({ history, match }: Props) {
  const [inputs, setInputs] = useState();
  const [errors, setErrors] = useState({});
  const [name, setName] = useState("");
  const cancelRequestRef = React.useRef();

  const definitionsUrl = serviceUrl.getTemplates();
  const policyUrl = serviceUrl.getPolicy({ policyId: match.params.policyId });
  const validateInfoUrl = serviceUrl.getValidateInfo({ policyId: match.params.policyId });

  const { data: definitionsData, isLoading: definitionsIsLoading, error: definitionsError } = useQuery({
    queryKey: definitionsUrl,
    queryFn: resolver.query(definitionsUrl),
  });
  const { data: policyData, isLoading: policyIsLoading, error: policyError } = useQuery({
    queryKey: policyUrl,
    queryFn: resolver.query(policyUrl),
  });
  const { data: validateInfoData, isLoading: validateInfoIsLoading, error: validateInfoError } = useQuery({
    queryKey: validateInfoUrl,
    queryFn: resolver.query(validateInfoUrl),
  });

  /**
   * Transform the policy object into shape that can be read by the child form
   * @param {object} policy - policy to read in and create input state from
   * @param {array} definitionsData - definitionsData referenced in policy
   * @returns {object} - new state object for "inputs" key
   */
  const formatPolicyDataForForm = useCallback((policyData, definitionsData) => {
    const newInputsState = {};
    policyData.definitions.forEach((definition: any) => {
      const policyDefinition = definitionsData.find(
        (policyDefinition: any) => policyDefinition.id === definition.policyTemplateId
      );
      // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
      newInputsState[policyDefinition.key] = {};
      // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
      const definitionRows = newInputsState[policyDefinition.key];
      definition.rules.forEach((rule: any) => {
        definitionRows[uuid.v4()] = rule;
      });
    });
    return newInputsState;
  }, []);

  useEffect(() => {
    if (policyData && definitionsData) {
      // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '{}' is not assignable to paramet... Remove this comment to see the full error message
      setInputs(formatPolicyDataForForm(policyData, definitionsData));
    }
  }, [formatPolicyDataForForm, policyData, definitionsData]);

  const [updatePolicyMutation, { isLoading: isUpdating }] = useMutation(
    (args) => {
      const { promise, cancel } = resolver.patchUpdatePolicy(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryCache.invalidateQueries(definitionsUrl),
    }
  );

  const [deletePolicyMutation, { isLoading: isDeleting }] = useMutation(
    (args) => {
      const { promise, cancel } = resolver.deletePolicy(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryCache.invalidateQueries(definitionsUrl),
    }
  );

  const updatePolicy = async () => {
    let policyObject = {
      // @ts-expect-error ts-migrate(2571) FIXME: Object is of type 'unknown'.
      id: policyData.id,
      name: name,
      // @ts-expect-error ts-migrate(2571) FIXME: Object is of type 'unknown'.
      teamId: policyData.teamId,
      definitions: [],
    };

    // @ts-expect-error ts-migrate(2571) FIXME: Object is of type 'unknown'.
    definitionsData.forEach((definition: any) => {
      let newDefinition = {
        policyTemplateId: definition.id,
      };
      let rules = [];
      // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
      const definitionRows = inputs[definition.key];
      for (let row in definitionRows) {
        rules.push(definitionRows[row]);
      }
      // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
      newDefinition["rules"] = rules;
      // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '{ policyTemplateId: any; }' is n... Remove this comment to see the full error message
      policyObject.definitions.push(newDefinition);
    });

    try {
      // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '{ body: { id: any; name: string;... Remove this comment to see the full error message
      await updatePolicyMutation({ body: policyObject, policyId: policyData.id });
      // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
      notify(<ToastNotification kind="success" title="Policy Updated" subtitle="Policy successfully updated" />);
      navigateBack();
    } catch (e) {
      // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
      notify(<ToastNotification kind="error" title="Something's Wrong" subtitle="Request to update policy failed" />);
    }
  };

  const deletePolicy = async () => {
    try {
      // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '{ policyId: any; }' is not assig... Remove this comment to see the full error message
      await deletePolicyMutation({ policyId: policyData.id });
      notify(
        // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
        <ToastNotification
          kind="success"
          title="Policy Deleted"
          // @ts-expect-error ts-migrate(2571) FIXME: Object is of type 'unknown'.
          subtitle={`Policy ${policyData.name} successfully deleted`}
        />
      );
      navigateBack();
    } catch (err) {
      const { data } = err && err.response;
      // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
      notify(<ToastNotification kind="error" title={`${data.status} - ${data.error}`} subtitle={data.message} />, {
        autoClose: 5000,
      });
    }
  };

  // State updates
  const setError = (error: any) => {
    setErrors((prevState) => ({ ...prevState, ...error }));
  };

  const setInput = async ({
    event: e,
    definitionKey,
    uuid
  }: any) => {
    const { name, value } = e.target;
    await setInputs((prevState) => {
      // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
      const prevStateDefinitionRows = prevState[definitionKey] ? prevState[definitionKey][uuid] : {};
      return {
        // @ts-expect-error ts-migrate(2698) FIXME: Spread types may only be created from object types... Remove this comment to see the full error message
        ...prevState,
        [definitionKey]: {
          ...prevState[definitionKey],
          [uuid]: { ...prevStateDefinitionRows, [name]: value },
        },
      };
    });
    validateRow(definitionKey);
  };

  const removeRow = async ({
    definitionKey,
    uuid
  }: any) => {
    // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
    let definitionRows = { ...inputs[definitionKey] };
    if (definitionRows) {
      delete definitionRows[uuid];
      await setInputs((prevState) => ({
        // @ts-expect-error ts-migrate(2698) FIXME: Spread types may only be created from object types... Remove this comment to see the full error message
        ...prevState,
        [definitionKey]: definitionRows,
      }));
      validateRow(definitionKey);
    }
  };

  /**
   *
   * @param {definitionKey} - key reference to a definition type e.g. static_code_analysis
   */
  const validateRow = (definitionKey: any) => {
    // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
    const definitionRows = inputs[definitionKey] || {};
    const definitionRowsInputCount = Object.keys(definitionRows).reduce((accum, uuid) => {
      const inputCount = Object.values(definitionRows[uuid]).filter(Boolean).length;
      accum += inputCount;
      return accum;
    }, 0);

    // Each row should have the same number of inputs as the number of inputs in the policy definition rules
    // @ts-expect-error ts-migrate(2571) FIXME: Object is of type 'unknown'.
    const matchingDefintion = definitionsData.find((definition: any) => definition.key === definitionKey);
    const isInvalid = Object.keys(definitionRows).length * matchingDefintion.rules.length !== definitionRowsInputCount;
    setErrors((prevState) => ({ ...prevState, [definitionKey]: isInvalid }));
  };

  useEffect(() => {
    return function cleanup() {
      queryCache.removeQueries(policyUrl, { exact: true });
    };
  }, [policyUrl]);
  // Local methods

  const navigateBack = () => {
    history.push(appLink.teamOverview({ teamId: match.params.teamId }));
  };
  const hasError = definitionsError || policyError || validateInfoError;
  const isLoading = definitionsIsLoading || policyIsLoading || validateInfoIsLoading;

  const form = {
    name,
    inputs,
    errors,
    setName,
    setInput,
    setError,
    removeRow,
    validateRow,
    onCancel: cancelRequestRef.current,
    affirmativeAction: updatePolicy,
    deletePolicy,
    isPerformingAffirmativeAction: isUpdating,
    isDeleting,
  };

  const helmetTitle = policyData?.name ? `${policyData.name} - ` : "";
  return (
    // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
    <div className={styles.container}>
      <Helmet>
        <title>{`${helmetTitle}Bosun Policies`}</title>
      </Helmet>
      <CreateEditPolicyHeader
        form={form}
        navigateBack={navigateBack}
        policy={policyData}
        type={POLICY_INTERACTION_TYPES.EDIT}
        validateInfo={validateInfoData}
        isLoading={isLoading}
        hasError={hasError}
      />
      {isLoading ? (
        <div className={styles.skeletonsContainer}>
          <TextInputSkeleton className={styles.textInputSkeleton} />
          <DefinitionSkeleton />
          <DefinitionSkeleton />
          <DefinitionSkeleton />
        </div>
      ) : hasError ? (
        <div style={{ marginTop: "2rem" }}>
          <ErrorMessage />
        </div>
      ) : (
        <CreateEditPolicyForm form={form} definitions={definitionsData} />
      )}
    </div>
  );
}

export default EditPolicy;
