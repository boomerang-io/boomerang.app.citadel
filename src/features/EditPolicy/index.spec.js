import React from "react";
import axios from "axios";
import { fireEvent, waitForElement } from "react-testing-library";
import EditPolicy from ".";
import MockAdapter from "axios-mock-adapter";
import { SERVICE_PRODUCT_DEFINITIONS_PATH, SERVICE_PRODUCT_POLICIES_PATH } from "Config/servicesConfig";

const route = "/111/policy/edit/222";
const props = {
  match: { params: { policyId: "222" } }
};

const definitions = [
  {
    id: "5cd49777f6ea74a9bb6ac629",
    key: "static_code_analysis",
    name: "Static Code Analysis",
    description: "The following policy metrics are validated from SonarQube data collected in your CI pipeline.",
    order: 0,
    config: [
      {
        key: "metric",
        label: "Metric",
        type: "select",
        defaultValue: "",
        required: false,
        description: "",
        options: [
          "issues-total",
          "issues-blocker",
          "issues-critical",
          "issues-major",
          "issues-minor",
          "issues-info",
          "issues-filesAnalyzed",
          "ncloc",
          "complexity",
          "violations"
        ]
      },
      {
        key: "operator",
        label: "Operator",
        type: "select",
        defaultValue: "",
        required: false,
        description: "",
        options: ["equal", "not equal", "less than", "less than or equal", "greater than", "greater than or equal"]
      },
      {
        key: "value",
        label: "Value",
        type: "text",
        defaultValue: "",
        required: false,
        description: "",
        options: null
      }
    ]
  }
];

const policy = {
  id: "5cd49adff6ea74a9bb6adef3",
  createdDate: "2019-06-21T00:00:00.000+0000",
  name: "Tyson's Policy",
  teamId: "5a8b331e262a70306622df73",
  definitions: [
    {
      ciPolicyDefinitionId: "5cd49777f6ea74a9bb6ac629",
      rules: [
        {
          metric: "lines",
          operator: "",
          value: "10"
        },
        {
          metric: "complexity",
          operator: "",
          value: "1000"
        }
      ]
    }
  ],
  stages: ["dev"]
};

describe("EditPolicy --- Snapshot", () => {
  beforeEach(() => {
    document.body.setAttribute("id", "app");
  });

  const mockAxios = new MockAdapter(axios);
  mockAxios.onGet(SERVICE_PRODUCT_DEFINITIONS_PATH).reply(200, definitions);
  mockAxios.onGet(SERVICE_PRODUCT_POLICIES_PATH + "/222").reply(200, policy);
  it("+++ renders correctly", async () => {
    const { baseElement, getByText } = renderWithRouter(<EditPolicy {...props} />, { route });
    await waitForElement(() => getByText(/Edit Policy Definitions/i));
    expect(baseElement).toMatchSnapshot();
  });
});

describe("EditPolicy --- RTL", () => {
  beforeEach(() => {
    document.body.setAttribute("id", "app");
  });

  test("renders error message when fetching definitions failed", async () => {
    const mockAxios = new MockAdapter(axios);
    mockAxios.onGet(SERVICE_PRODUCT_DEFINITIONS_PATH).reply(404);
    mockAxios.onGet(SERVICE_PRODUCT_POLICIES_PATH + "/222").reply(200, policy);

    const { getByText } = renderWithRouter(<EditPolicy {...props} />, { route });
    const errorMessage = await waitForElement(() => getByText("Don’t lose your daks"));
    expect(errorMessage).toBeInTheDocument();
  });

  test("renders error message when fetching policy failed", async () => {
    const mockAxios = new MockAdapter(axios);
    mockAxios.onGet(SERVICE_PRODUCT_DEFINITIONS_PATH).reply(200, definitions);
    mockAxios.onGet(SERVICE_PRODUCT_POLICIES_PATH + "/222").reply(404);

    const { getByText } = renderWithRouter(<EditPolicy {...props} />, { route });
    const errorMessage = await waitForElement(() => getByText("Don’t lose your daks"));
    expect(errorMessage).toBeInTheDocument();
  });

  test("it is able to update policy only if it has a name", async () => {
    const mockAxios = new MockAdapter(axios);
    mockAxios.onGet(SERVICE_PRODUCT_DEFINITIONS_PATH).reply(200, definitions);
    mockAxios.onGet(SERVICE_PRODUCT_POLICIES_PATH + "/222").reply(200, policy);

    const { getByPlaceholderText, getByTestId } = renderWithRouter(<EditPolicy {...props} />, { route });

    const saveButton = await waitForElement(() => getByTestId("policy-header-affirmative-action"));

    expect(saveButton).toBeEnabled();

    const nameInput = getByPlaceholderText(/name/i);
    fireEvent.change(nameInput, { target: { value: "" } });
    expect(saveButton).toBeDisabled();

    fireEvent.change(nameInput, { target: { value: "test" } });
    expect(saveButton).toBeEnabled();
  });

  test("save button is disabled while saving", async () => {
    const mockAxios = new MockAdapter(axios);
    mockAxios.onGet(SERVICE_PRODUCT_DEFINITIONS_PATH).reply(200, definitions);
    mockAxios.onGet(SERVICE_PRODUCT_POLICIES_PATH + "/222").reply(200, policy);

    const { getByPlaceholderText, getByTestId } = renderWithRouter(<EditPolicy {...props} />, { route });

    const saveButton = await waitForElement(() => getByTestId("policy-header-affirmative-action"));

    const nameInput = getByPlaceholderText(/name/i);
    fireEvent.change(nameInput, { target: { value: "test" } });
    expect(saveButton).toBeEnabled();

    fireEvent.click(saveButton);
    expect(saveButton).toBeDisabled();
  });

  test("delete modals appears and delete button is disabled while deleting", async () => {
    const mockAxios = new MockAdapter(axios);
    mockAxios.onGet(SERVICE_PRODUCT_DEFINITIONS_PATH).reply(200, definitions);
    mockAxios.onGet(SERVICE_PRODUCT_POLICIES_PATH + "/222").reply(200, policy);

    const { getByText } = renderWithRouter(<EditPolicy {...props} />, { route });

    const deleteButton = await waitForElement(() => getByText(/delete/i));
    expect(deleteButton).toBeEnabled();

    fireEvent.click(deleteButton);
    getByText(/it will be gone/i);

    fireEvent.click(getByText(/yes/i));
    expect(deleteButton).toBeDisabled();
  });
});
