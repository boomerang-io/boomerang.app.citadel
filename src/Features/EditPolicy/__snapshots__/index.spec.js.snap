// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPolicy --- Snapshot +++ renders correctly 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <div
        class="container"
      >
        <div
          class="content"
        >
          <div
            class="info"
          >
            <h1
              class="title"
            >
              Edit Policy
            </h1>
            <div
              class="metadataContainer"
            >
              <p
                class="metadata"
              >
                <span
                  class="metaDataLabel"
                >
                  Created: 
                </span>
                 
                <time>
                  January 1st, 2019
                </time>
              </p>
              <button
                class="validationInfoButton"
              >
                How do I use this?
              </button>
            </div>
          </div>
          <section
            class="buttons"
          >
            <button
              class="button bx--btn bx--btn--field bx--btn--secondary"
              tabindex="0"
              type="button"
            >
              Cancel
            </button>
            <button
              class="button bx--btn bx--btn--field bx--btn--danger"
              tabindex="0"
              type="button"
            >
              Delete
              <svg
                aria-hidden="true"
                aria-label="Delete"
                class="bx--btn__icon"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 6h1v6H6zm3 0h1v6H9z"
                />
                <path
                  d="M2 3v1h1v10c0 .6.4 1 1 1h8c.6 0 1-.4 1-1V4h1V3H2zm2 11V4h8v10H4zM6 1h4v1H6z"
                />
              </svg>
            </button>
            <button
              class="button bx--btn bx--btn--field bx--btn--primary"
              data-testid="policy-header-affirmative-action"
              tabindex="0"
              type="submit"
            >
              Save
              <svg
                aria-hidden="true"
                aria-label="Edit"
                class="bx--btn__icon"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.9 4.6l-2.5-2.5c-.1 0-.3-.1-.4-.1H3c-.6 0-1 .4-1 1v10c0 .6.4 1 1 1h10c.6 0 1-.4 1-1V5c0-.1-.1-.3-.1-.4zM6 3h4v2H6V3zm4 10H6V9h4v4zm1 0V9c0-.6-.4-1-1-1H6c-.6 0-1 .4-1 1v4H3V3h2v2c0 .6.4 1 1 1h4c.6 0 1-.4 1-1V3.2l2 2V13h-2z"
                />
              </svg>
            </button>
          </section>
        </div>
        <div
          class="bx--modal"
          role="presentation"
          tabindex="-1"
        >
          <div
            class="bx--modal-container"
          >
            <div
              class="bx--modal-header"
            >
              <p
                class="bx--modal-header__label bx--type-delta"
              >
                How to use this policy
              </p>
              <p
                class="bx--modal-header__heading bx--type-beta"
              >
                Validation Guide
              </p>
              <button
                class="bx--modal-close"
                title="Close the modal"
                type="button"
              >
                <svg
                  aria-label="Close the modal"
                  class="bx--modal-close__icon"
                  focusable="false"
                  height="20"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="bx--modal-content"
            >
              <h2
                class="modalSectionTitle"
              >
                Validation Endpoint
              </h2>
              <p>
                undefined/bosun/validate
              </p>
              <h2
                class="modalSectionTitle"
              >
                Sample Payload
              </h2>
              <div
                class="bx--snippet bx--snippet--multi"
              >
                <div
                  aria-label="code-snippet"
                  class="bx--snippet-container"
                  role="textbox"
                  tabindex="0"
                >
                  <code>
                    <pre>
                      {
 "policyId": "5db9a8c7b01c530001b838d1",
 "referenceId": null,
 "referenceLink": null,
 "labels": {
  "artifact-path": "",
  "artifact-name": "",
  "artifact-version": "",
  "sonarqube-id": "",
  "sonarqube-version": ""
 },
 "annotations": null,
 "data": null
}
                    </pre>
                  </code>
                </div>
                <button
                  aria-label="Copy annotations to clipboard"
                  class="bx--snippet-button"
                  title="Copy annotations to clipboard"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="bx--snippet__icon"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    style="will-change: transform;"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M14 5v9H5V5h9m0-1H5a1 1 0 0 0-1 1v9a1 1 0 0 0 1 1h9a1 1 0 0 0 1-1V5a1 1 0 0 0-1-1z"
                    />
                    <path
                      d="M2 9H1V2a1 1 0 0 1 1-1h7v1H2z"
                    />
                  </svg>
                  <div
                    class="bx--btn--copy__feedback"
                    data-feedback="Copied!"
                  />
                </button>
              </div>
              <h2
                class="modalSectionTitle"
              >
                Next Steps (things to adjust)
              </h2>
              <ol
                class="bx--list--ordered"
              >
                <li
                  class="bx--list__item"
                >
                  Update 
                  <code>
                    ReferenceId
                  </code>
                   to a unique identifier to ensure you can track specific xyz
                </li>
                <li
                  class="bx--list__item"
                >
                  Update 
                  <code>
                    ReferenceLink
                  </code>
                   a link back to xyz
                </li>
                <li
                  class="bx--list__item"
                >
                  Supply
                   
                  <a
                    href="https://github.com/boomerang-io/boomerang.docs/blob/stable/content/bosun.md#labels-required"
                  >
                    labels
                  </a>
                   
                  that can be used by the integrations to integrate and retrieve backend information
                </li>
                <li
                  class="bx--list__item"
                >
                  Call the validation endpoint with the updated data
                </li>
              </ol>
            </div>
            <div
              class="bx--modal-footer"
              inputref="[object Object]"
            >
              <button
                class="bx--btn bx--btn--primary"
                tabindex="0"
                type="button"
              >
                Close
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="content"
        >
          <div
            class="bx--bmrg-text-input"
          >
            <div
              class="bx--form-item bx--text-input-wrapper"
            >
              <label
                class="bx--label"
                for="name"
              >
                <div
                  class="bx--bmrg-text-input__label"
                >
                  <div>
                    Name
                  </div>
                </div>
              </label>
              <div
                class="bx--text-input__field-wrapper"
              >
                <input
                  class="bx--text-input bx--text__input"
                  id="name"
                  name="name"
                  placeholder="Name"
                  type="text"
                  value="Tyson's Policy"
                />
              </div>
            </div>
          </div>
          <section
            class="section"
          >
            <h2>
              Package Safe List
            </h2>
            <p>
              The following package and artifact validation is validated against scans by JFrog Xray. Artifact and Version are validated through regular expression.
            </p>
            <div>
              <div
                class="row"
              >
                <div
                  class="bx--bmrg-text-input"
                >
                  <div
                    class="bx--form-item bx--text-input-wrapper"
                  >
                    <label
                      class="bx--label"
                      for="artifact"
                    >
                      <div
                        class="bx--bmrg-text-input__label"
                      >
                        <div>
                          Package
                        </div>
                      </div>
                    </label>
                    <div
                      class="bx--text-input__field-wrapper"
                    >
                      <input
                        autocomplete="off"
                        class="bx--text-input bx--text__input"
                        id="artifact"
                        name="artifact"
                        placeholder="Package"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <button
                  class="delete"
                >
                  <svg
                    aria-hidden="true"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    style="will-change: transform;"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 6h1v6H6zm3 0h1v6H9z"
                    />
                    <path
                      d="M2 3v1h1v10c0 .6.4 1 1 1h8c.6 0 1-.4 1-1V4h1V3H2zm2 11V4h8v10H4zM6 1h4v1H6z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="row"
              >
                <div
                  class="bx--bmrg-text-input"
                >
                  <div
                    class="bx--form-item bx--text-input-wrapper"
                  >
                    <label
                      class="bx--label"
                      for="artifact"
                    >
                      <div
                        class="bx--bmrg-text-input__label"
                      >
                        <div>
                          Package
                        </div>
                      </div>
                    </label>
                    <div
                      class="bx--text-input__field-wrapper"
                    >
                      <input
                        autocomplete="off"
                        class="bx--text-input bx--text__input"
                        id="artifact"
                        name="artifact"
                        placeholder="Package"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <button
                  class="delete"
                >
                  <svg
                    aria-hidden="true"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    style="will-change: transform;"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 6h1v6H6zm3 0h1v6H9z"
                    />
                    <path
                      d="M2 3v1h1v10c0 .6.4 1 1 1h8c.6 0 1-.4 1-1V4h1V3H2zm2 11V4h8v10H4zM6 1h4v1H6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <button
              class="bx--btn bx--btn--field bx--btn--primary"
              tabindex="0"
              type="button"
            >
              Add Rule
              <svg
                aria-hidden="true"
                aria-label="Add Rule"
                class="bx--btn__icon"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 7V3H7v4H3v2h4v4h2V9h4V7z"
                />
              </svg>
            </button>
          </section>
        </div>
        <div
          class="containerGraphics"
        >
          <svg
            class="graphic"
            data-name="Layer 1"
            id="prefix__Layer_1"
            viewBox="0 0 202.29 228.51"
          >
            <defs>
              <style>
                .prefix__cls-2{fill:#20d5d2}.prefix__cls-3{fill:none;stroke:#323c3c;stroke-linecap:round;stroke-miterlimit:10;stroke-width:5px}
              </style>
            </defs>
            <title>
              Artboard 54
            </title>
            <path
              d="M179.68 167.15a5.32 5.32 0 014 8.8l-24.07 27.79a5.34 5.34 0 01-4 1.83H22.31a5.32 5.32 0 01-3.41-9.41l33.1-27.5a5.28 5.28 0 013.4-1.23l65.52-.15c1.32 5.69 7.52 16.94 7.52 16.94s7-11.47 8.4-17z"
              fill="#f2f4f8"
            />
            <path
              class="prefix__cls-2"
              d="M98.2 44.67a27.15 27.15 0 114.12 24.71 27.15 27.15 0 01-4.12-24.71zm39.16 12.74a14 14 0 10-17.69 9 14.05 14.05 0 0017.69-9zM130.88 110.1l-.46 13.28-52.97-17.24 7.46-11 45.97 14.96zM129.25 157.41a8.21 8.21 0 017.8 8.37 7.5 7.5 0 01-.23 1.46H120.9a7.9 7.9 0 01-.24-1.88 8.17 8.17 0 018.38-8zM128.42 184.22s-6.2-11.25-7.52-16.94h15.92c-1.37 5.47-8.4 16.94-8.4 16.94zM46.91 135.77a8.21 8.21 0 0111.4-2.07l.33.25a8.19 8.19 0 011.74 11.17C57.8 148.83 43 155.75 43 155.75s1.32-16.26 3.91-19.98z"
            />
            <path
              class="prefix__cls-3"
              d="M120.9 167.28l-65.52.15a5.28 5.28 0 00-3.4 1.23l-33.08 27.5a5.32 5.32 0 003.41 9.41h133.31a5.34 5.34 0 004-1.83L183.71 176a5.32 5.32 0 00-4-8.8l-42.86.09M132.41 27.25A27.15 27.15 0 1198.2 44.67a27.15 27.15 0 0134.21-17.42zM137.03 13.07l-4.62 14.18"
            />
            <path
              class="prefix__cls-3"
              d="M102.32 69.38L84.91 95.14l-7.46 11-18.81 27.81M131.95 79.02l-1.07 31.08-.46 13.28-1.17 34.03M84.91 95.13l45.96 14.97M77.45 106.14l52.97 17.24M137.36 57.41a14 14 0 11-9-17.69 14.05 14.05 0 019 17.69zM46.91 135.77a8.21 8.21 0 0111.4-2.07l.33.25a8.19 8.19 0 011.74 11.17C57.8 148.83 43 155.75 43 155.75s1.32-16.26 3.91-19.98z"
            />
            <path
              class="prefix__cls-3"
              d="M129.25 157.41H129a8.17 8.17 0 00-8.38 8 7.9 7.9 0 00.24 1.88c1.32 5.69 7.52 16.94 7.52 16.94s7-11.47 8.4-17a7.5 7.5 0 00.23-1.46 8.21 8.21 0 00-7.76-8.36z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</body>
`;
